### PR TITLE
close #11330 sets uses optimized countSetBits

### LINKED ISSUE #11330: sets use non-optimized countBits32
countbits32

/cc @krux02 

related #10617

### PR DESCRIPTION
**Pull request by ringabout**
close #11330

followup
- [ ] now that sets is an import module, move it to lib/std/private/sets_builtin
https://github.com/nim-lang/Nim/pull/17334#discussion_r598067587

- [ ] ditto for countbits_impl

### CODE REVIEW

File: lib/system/sets.nim [hunk 1]
```diff
@@ -8,26 +8,160 @@
...
+const noUndefined {.used.} = defined(noUndefinedBitOpts)
+const useGCC_builtins = (defined(gcc) or defined(llvm_gcc) or
+                         defined(clang)) and useBuiltins
+const useICC_builtins = defined(icc) and useBuiltins
+const useVCC_builtins = defined(vcc) and useBuiltins
 
 type
   NimSet = array[0..4*2048-1, uint8]
 
 # bitops can't be imported here, therefore the code duplication.
```
Comment (timotheecour) [inline lib/system/sets.nim]: is this comment about `code duplication` still relevant? which code duplication?

Comment (ringabout) [reply lib/system/sets.nim]: make sense

Comment (ringabout) [reply lib/system/sets.nim]: fixed

File: lib/system.nim [hunk 2]
```diff
@@ -2339,7 +2339,7 @@ when notJSnotNims:
   when hostOS != "standalone" and hostOS != "any":
     include "system/dyncalls"
 
-  include "system/sets"
+  from system/sets import cardSet
```
Comment (Araq) [inline lib/system.nim]: This is blocking the merge until my basic IC tests are merged. Yet another system module dependency...

Comment (ringabout) [reply lib/system.nim]: ok

File: lib/pure/bitops.nim [hunk 3]
```diff
@@ -27,6 +27,13 @@
 
 import macros
 import std/private/since
+from std/private/vmutils import forwardImpl, toUnsigned
+
+
+# sets.nim cannot import bitops, but bitops can use include
+# system/sets to eliminate code duplication. sets.nim defines
+# countBits32 and countBits64.
+import system/sets
```
Comment (timotheecour) [inline lib/pure/bitops.nim]: * now that sets is an import module, move it to `lib/std/private/sets_builtin`

this way, we also avoid 2 import modules with same name:
lib/pure/collections/sets.nim
system/sets

which can cause issues (eg for docs https://nim-lang.github.io/Nim/sets.html, references etc)

* this comment also needs to be upgrade to reflect this and fact that it's import, not include

File: lib/system/countbits_impl.nim [hunk 4]
```diff
@@ -9,17 +9,91 @@
...
+  # Count the number of bits set to 1 in an integer a, and return that count in dst.
+  func builtin_popcnt32(a: cint): cint {.
+      importc: "_popcnt", header: "<immintrin.h>".}
+  func builtin_popcnt64(a: uint64): cint {.
+      importc: "_popcnt64", header: "<immintrin.h>".}
+
+
+func countSetBitsImpl*(x: SomeInteger): int {.inline.} =
+  ## Counts the set bits in an integer (also called `Hamming weight`:idx:).
+  runnableExamples:
```
Comment (timotheecour) [inline lib/system/countbits_impl.nim]: move runnableExamples to bitops.countSetBits

File: lib/system/countbits_impl.nim [hunk 5]
```diff
@@ -9,17 +9,91 @@
...
+  # Counts the number of one bits (population count) in a 16-, 32-, or 64-byte unsigned integer.
+  func builtin_popcnt16(a2: uint16): uint16 {.
+      importc: "__popcnt16", header: "<intrin.h>".}
+  func builtin_popcnt32(a2: uint32): uint32 {.
+      importc: "__popcnt", header: "<intrin.h>".}
+  func builtin_popcnt64(a2: uint64): uint64 {.
+      importc: "__popcnt64", header: "<intrin.h>".}
+
+elif useICC_builtins:
+
```
Comment (timotheecour) [inline lib/system/countbits_impl.nim]:
```suggestion
```

**Review by timotheecour — APPROVED** [review]
LGTM after addressing comments

File: lib/pure/bitops.nim (hunk 3 above)
Comment (konsumlamm) [reply lib/pure/bitops.nim]: For consistency with `channels_builtin` (#17330), perhaps use `sets_builtin`?

Comment (timotheecour) [reply lib/pure/bitops.nim]: sure, i updated the comment

Comment (ringabout) [reply lib/pure/bitops.nim]: `countbits_impl` can also be moved to `std/private`, so I left it the following PR.

File: lib/system/countbits_impl.nim (hunk 4 above)
Comment (ringabout) [reply lib/system/countbits_impl.nim]: fixed

**Review by timotheecour — APPROVED** [review]

File: lib/pure/bitops.nim [hunk 6]
```diff
@@ -27,6 +27,13 @@
 
 import macros
 import std/private/since
+from std/private/vmutils import forwardImpl, toUnsigned
+
+
+# sets.nim cannot import bitops, but bitops can use import
+# system/sets to eliminate code duplication. sets.nim defines
+# countBits32 and countBits64.
+import system/sets
```
Comment (Araq) [inline lib/pure/bitops.nim:36]:
```suggestion
```

Comment (Araq) [inline lib/pure/bitops.nim]:
```suggestion
```

File: lib/system.nim [hunk 7]
```diff
@@ -2339,8 +2339,7 @@ when notJSnotNims:
   when hostOS != "standalone" and hostOS != "any":
     include "system/dyncalls"
 
-  import system/countbits_impl
-  include "system/sets"
+  from system/sets import cardSet
```
Comment (Araq) [inline lib/system.nim]: Undo this change please, `system/sets` is now only an include file

Comment (ringabout) [reply lib/system.nim]: OK, will finish it later

File: lib/system/sets.nim [hunk 8]
```diff
@@ -8,13 +8,13 @@
...
 # set handling
+import system/countbits_impl
 
 type
   NimSet = array[0..4*2048-1, uint8]
 
-# bitops can't be imported here, therefore the code duplication.
 
-proc cardSet(s: NimSet, len: int): int {.compilerproc, inline.} =
+proc cardSet*(s: NimSet, len: int): int {.compilerproc, inline.} =
```
Comment (ringabout) [inline lib/system/sets.nim]:
```suggestion
proc cardSet(s: NimSet, len: int): int {.compilerproc, inline.} =
```

File: lib/system/sets.nim [hunk 9]
```diff
@@ -8,11 +8,11 @@
 #
 
 # set handling
+import system/countbits_impl
```
Comment (ringabout) [inline lib/system/sets.nim]:
```suggestion
```

File: lib/system.nim [hunk 10]
```diff
@@ -2339,8 +2339,7 @@ when notJSnotNims:
   when hostOS != "standalone" and hostOS != "any":
     include "system/dyncalls"
 
-  import system/countbits_impl
-  include "system/sets"
+  include system/sets
```
Comment (ringabout) [inline lib/system.nim]:
```suggestion
  import system/countbits_impl
  include "system/sets"
```

**Pull request merged**